process/block/metablock.go

```diff
@@ -2073,6 +2073,12 @@ func (mp *metaProcessor) receivedShardHeader(headerHandler data.HeaderHandler, s
 		"hash", shardHeaderHash,
 	)
 
+	if mp.isWaitingForNextHeader.IsSet() {
```

Comment:
there is no guarantee that the received header is the one we requested or it is in fact another one, so in the second case we would falsely exit the wait.